docs: Update section on rendering 📚 (#99)
ratatui-org/ratatui-website: src/concepts/rendering-under-the-hood.md

# How does Ratatui work?

You may have read in previous sections that Ratatui is a immediate mode rendering library. But what
does that really mean? And how is it implemented? In this section, we are going to discuss what it

> means "to be a `Widget`" and what happens behind the scenes in Ratatui.

Reviewer: ```suggestion
does that really mean? And how is it implemented? In this section, we will discuss how Ratatui renders `Widget`
to the screen, starting with the `Terminal`'s `draw` method and ending with your chosen backend library.
```

In Ratatui, the primary mechanism for making something that is renderable is through the `Widget`

> trait.

Reviewer: Perhaps kick off from the top of the user's interaction like so and introduce the map of the discussion:

```suggestion
Rendering an application's UI, starts with the application calling the [`draw` ] method on an instance of the [`Terminal`] struct. This method takes a closure[^closure] which accepts an instance of a `Frame`, which in turn has a reference to the current buffer which can be rendered to. Inside the `draw` method, applications can call [`Frame::render_widget()`] or [`Frame::render_stateful_widget()`] to render the state of each widget to certain locations within the ...

Talk about how each call to draw represents a single frame, and should contain all the details (doesn't need to be an admonish).

[^closure]: some footnote link to beginner level docs about closures (or make it a real link in text)
[`Terminal`]: https://docs.rs/ratatui/...
[`draw`]: ...
[`Frame`]: ...

## Widgets

Ratatui's 

Ratatui provides a [`Widget`] trait that enables any implementer to render something to the terminal. The `Widget` trait is implemented for a selection internal types (like `Paragraph`, `Table` and `List`, as well as in external crates like XXXExample, .... Applications can also create their own widgets to represent the rendering of the application. The trait provides a single `render` method which takes an area and a mutable reference to the buffer to render to. The `render` method is expected to draw the current state of the widget within the bounds of the provided area by calling methods on the buffer to insert styled text, or directly manipulate each individual [`Cell`]. The `Cell` is the smallest ...
```

Then a smallish hello world example - no main, just the terminal `draw` call, frame, Paragraph, etc.

Then roll through that example by talking specifics of how each part fits together line by line and action by action.

    );
}
```

This `Frame` struct contains a method called `render_widget` which takes any object that implements
the `Widget` trait. This `render_widget` method calls the `Widget::render` method on the type-erased
`Widget` struct.

```rust
    pub fn render_widget<W>(&mut self, widget: W, area: Rect)
    where
        W: Widget,
    {
        widget.render(area, self.buffer);
    }
```

As mentioned above, implementing the `Widget` trait is the primary way to make a struct renderable.
As an example, here's the full implementation for the `Clear` widget:

```rust
pub struct Clear;

impl Widget for Clear {
    fn render(self, area: Rect, buf: &mut Buffer) {
        for x in area.left()..area.right() {
            for y in area.top()..area.bottom() {
                buf.get_mut(x, y).reset();
            }
        }
    }
}
```

There are 2 things that you should notice.

1. The `Widget` gets the current area as a `Rect`.
2. The `Widget` gets a mutable reference to a "`Buffer`".

The top level widget usually gets a reference to the entire size of the terminal as a `Rect`, i.e.
`f.size()`. However, nested widgets may get a smaller area. Any implementation of a `Widget` must

> only draw "within the lines" of the `Rect` area.

Reviewer: Perhaps phrase this more as the frame has a `size()` method that returns a...

Rationale: in the "how rendering works", we don't want to go deep into "how to structure an app's widgets", so top level widget is meaningless - often apps won't have one at all.

Also, "must" is pretty strong here. If it's must then we should enforce it somehow in `frame.render_widget()`. It's only "should" right now.

You might think that a `Widget` renders directly to the terminal but that is not the case. Ratatui

> uses something called a "double-buffer" rendering technique.

Reviewer: I'd prefer to try to start sentences with the thing that something does, rather than what doesn't happen. Like:

```suggestion
Ratatui widgets render to an intermediate buffer before the information is rendered to the screen. This is in
contrast to using the terminal backends directly where writing text to the screen can occur at any time. The
process of creating a single Frame to render to the screen starts when the application calls ... (note active voice
rather than "is called"). The terminal creates a new Frame value that ...

After the application's closure runs, ...
```

Every time `terminal.draw(|f| ...)` is called, a new [`Frame`] struct is constructed that contains a
mutable reference to an instance of the `Buffer` struct:

```rust
pub struct Frame<'a> {
    // --snip--
    /// The buffer that is used to draw the current frame
    buffer: &'a mut Buffer,
}
```

The `Buffer` represents the area in which content is going to be drawn to the terminal. The `Buffer`

> struct implements a number of methods such as [`get_mut`] and [`set_string`]:

Reviewer: ```suggestion
The `Buffer` represents an area that the application can draw into by manipulating its contents. The `Buffer`
struct implements methods to write text, set styles on particular areas and manipulate individual cells. A cell is
...
```

```rust
pub fn get_mut(&mut self, x: u16, y: u16) -> &mut Cell {
}

pub fn set_string<S>(&mut self, x: u16, y: u16, string: S, style: Style)
where
    S: AsRef<str> {
}
```

These methods allow any implementation of the `Widget` trait to write into any parts of the
`Buffer`.

> A [`Buffer`] is essentially a collection of `Cell`s:

Reviewer: ```suggestion
A [`Buffer`] contains a `Cell` to represent the rows and columns of the terminal's display area (note this might not be the full terminal height and width because ...). The buffer has a content field that is a `Vec<Cell>` but widgets will generally interact with `...`
```

```rust
pub struct Buffer {
    // --snip--
    pub content: Vec<Cell>,
}
```

> In Ratatui, this `Cell` struct is the smallest renderable unit of code. A `Cell` essentially has a

Reviewer: ```suggestion
In Ratatui, this `Cell` struct is the smallest renderable unit of code. Each `Cell` tracks the symbol and style information, for the given row and column ...
```

`String` which is the symbol that will be drawn in that `Cell` as well as `Style` information
(foreground color, background color, modifiers etc).

```rust
// pseudo code
pub struct Cell {
  content: String,
  style: Style,
}
```

> ```admonish
> `Cell`s are the closest analog to a "pixel" in a terminal.
> Most terminals render monospaced text 2 high and 1 wide.
> A `Cell` in Ratatui should usually contain 1 wide string content.
> ```

Reviewer: drop the admonish?

```suggestion
`Cell`s are similar to a "pixel" in a graphical UI, bu instead of storing just a single point with color information
they store a single character. Terminals generally render text so that each individual cell takes up space
approximately twice as high as it is wide.
```

Add an admonish for Emoji (and perhaps link to a section)

Author: Skipping Emoji / unicode discussion for now until we add a separate page for it. Once I do, I'll update this page with a link to it.

> Here's a simplified class diagram for reference:

Reviewer: I'd drop this - too enterprisey. Can we go less formal style for this perhaps?

```mermaid
classDiagram
    class Cell {
        +content: String
        +style: Style
    }

    class Buffer {
        +area: Rect
        +content: Vec<Cell>
    }

    class Frame {
        +buffer: mut Buffer
        +render_widget()
    }

    class Widget {
        <<interface>>
        +render(area: Rect, buf: &mut Buffer)
    }

    class Paragraph
    class Block
    class Table

    Buffer --> Cell
    Frame --> Buffer
    Widget <|.. Paragraph
    Widget <|.. Block
    Widget <|.. Table

```

> Basically what this all means that any content rendered to a `Buffer` is only stored in that

Reviewer: ```suggestion
This all means that any content rendered to a `Buffer` is only stored in that
```

> particular `Buffer` struct in the form of `Cell`s that contain string and style information.

Reviewer: ```suggestion
`Buffer` that is attached to the frame during the `draw` call. The buffer stores the `Cell` values that contain symbol and style information.
```

> ```admonish note
> Because any styling has to happen though the `Style` struct (i.e. methods that apply
> `Style`s to `Cell`s) you can't use ansi escape sequences directly in the string content.
> 
> As a workaround, what you can do however is convert the ansi escape sequences into a
> `ratatui::text::Text` struct that contains the appropriate `Style`s. There's a crate for this that
> might help: <https://crates.io/crates/ansi-to-tui>. And this `Text` struct that you create can be
> passed into the `Paragraph` widget to get the desired effect.
> 
> Even if this particular crate doesn't work for you this approach is probably what you want to use if
> you want to render ansi escape sequences in Ratatui.
> You can learn more about the text related of Ratatui features and displaying text
> [here](./../how-to/render/display-text.md).
> ```

Reviewer: Make this shorter and link to a page:
```suggestion
```admonish note
ANSI Escape sequences for color and style that are stored in the cell's symbol are not rendered as the style information stored in the cell is used to style each character position. If your text has ANSI styling info, consider using the `ansi-to-tui` crate to convert this to a `Text` value before rendering. See ...
```
```

After all the `Widget`s are rendered to the same `Buffer` struct, i.e. after

> `terminal.draw(|f| ...)` returns, [`terminal.flush()`] is called. This is where the content is

Reviewer: After the closure provided to the `draw` method finishes, the `draw` method calls `Terminal::flush()`. This causes the buffer to be written to the screen. Ratatui uses a double buffer approach so that <REASON:  usually this is something like "the process of writing from the buffer to the screen happens all at once and the next buffer can be constructed while the first one is being rendered, but this isn't something that actually happens AFAIK>

actually written out to the terminal.

In `flush()`, the difference between the previous and the current buffer is obtained and that is
passed it to the current backend for drawing to your actual terminal.

```rust
    pub fn flush(&mut self) -> io::Result<()> {
        let previous_buffer = &self.buffers[1 - self.current];
        let current_buffer = &self.buffers[self.current];
        let updates = previous_buffer.diff(current_buffer);
        if let Some((col, row, _)) = updates.last() {
            self.last_known_cursor_pos = (*col, *row);
        }
        self.backend.draw(updates.into_iter())
    }

> ```

Reviewer: drop?

After `flush()`, the buffers are swapped and the next time `terminal.draw(|f| ...)` is called a
`Frame` is constructed with the other `Buffer` struct. This is why it is called a double-buffer
rendering technique.

The important thing to note is that because all widgets render to the same `Buffer` within a single
`terminal.draw(|f| ...)` call, rendering of different widgets may overwrite the same `Cell`, so the
order in which widgets are rendered is relevant. For example, in this `draw` example below,
`"content1"` will be overwritten by `"content2"` which will be overwritten by `"content3"` in the
`Buffer` struct, and Ratatui will only ever write out `"content3"` to the terminal:

```rust
terminal.draw(|f| {
    f.render_widget(Paragraph::new("content1"), f.size());
    f.render_widget(Paragraph::new("content2"), f.size());
    f.render_widget(Paragraph::new("content3"), f.size());
})
```

This was a sneak peak into what happens under the hood in Ratatui. If you have any questions, feel

> free to open issues or discussions on [the main repository](https://github.com/ratatui-org/ratatui).

Reviewer: In summary, the application calls draw to ..., and then blah blah blah. Each widget does blah blah blah.... Once the user inteface closure completes, ratatui ...

Perhaps drop the questions, etc. part (we don't need to repeat that on every page or if we do, then just having a common snippet would be better.

Author: https://github.com/ratatui-org/ratatui-book/pull/99/commits/ef16b8baf97267e8e4fb1e836599165995415beb
